### PR TITLE
Go 1.17

### PR DESCRIPTION
**Pull request by aykevl**
This is still a WIP, not all features are supported yet.

See #2006

### CODE REVIEW

**Comment by aykevl** [comment]
I've now also implemented the [three new language features](https://tip.golang.org/doc/go1.17#language) of Go 1.17.

**Comment by aykevl** [comment]
This should be ready once CircleCI publishes the 1.17 containers: https://hub.docker.com/r/circleci/golang/tags?page=1&ordering=last_updated&name=1.17-stretch

**Comment by dkegel-fastly** [comment]
Have you tried building with circleci's rc2 image, e.g. circleci/golang:1.17rc2-buster ?

**Review by niaow — APPROVED** [review]
This looks fairly complete, and should be good to go once circleci is ready

File: .circleci/config.yml
Comment (deadprogram) [inline .circleci/config.yml:382]: I am not sure that CircleCI actually plans to release a 1.17 version of this image. The reason is that they consider this image legacy. See https://circleci.com/developer/images/image/cimg/go

Comment (aykevl) [reply .circleci/config.yml:382]: The problem is that we rely on a specific Debian version. The cimg images seem to be based on Ubuntu (probably a recent version?). The version/variant of the underlying OS is important because it is the lowest that the TinyGo binaries will run on. For example, if we switched to Debian bullseye (the latest version), the prebuilt binaries of TinyGo wouldn't run on Debian buster anymore (the previous Debian version). Therefore, we need to use a base image that is as old as possible but still somewhat supported.

I'm thinking that we should probably create our own images with all software preinstalled, that should hopefully cut down on CI build time (no need to install anything) but more importantly provides stability (no packages that suddenly disappear from upstream). But perhaps not before this PR is merged.

In this case, the -rc2 image versions have been published so I'm hoping the final version will also be published soon.

Comment (deadprogram) [reply .circleci/config.yml:382]: I am not too sure on that. It has been a lot longer than previous releases, and the repo with the legacy images is now archived and read-only: https://github.com/CircleCI-Public/circleci-dockerfiles

In addition the LLVM 11 images seemed to have issues, which is why I merged https://github.com/tinygo-org/tinygo/commit/931f87f96a36565cf01291e9a306a808884e5b97 so that the GH action would stop failing on the *current* version.

Comment (deadprogram) [reply .circleci/config.yml:382]: Looks like it was finally pushed: https://hub.docker.com/layers/circleci/golang/1.17-stretch

**Comment by deadprogram** [comment]
CI now passing. Merging, thanks for working on this @aykevl

**Pull request merged**
